plugins/node/opentelemetry-instrumentation-aws-lambda/src/aws-lambda.ts

```diff
@@ -0,0 +1,216 @@
+      callback: Callback
+    ) {
+      const name = context.functionName;
+      const span = plugin.tracer.startSpan(name, {
+        kind: SpanKind.SERVER,
+        attributes: {
+          [FaasAttribute.FAAS_EXECUTION]: context.awsRequestId,
+          [FaasAttribute.FAAS_ID]: context.invokedFunctionArn,
+          [CLOUD_RESOURCE.ACCOUNT_ID]: AwsLambdaInstrumentation._extractAccountId(
+            context.invokedFunctionArn
+          ),
+        },
+      });
+
+      // Lambda seems to pass a callback even if handler is of Promise form, so we wrap all the time before calling
+      // the handler and see if the result is a Promise or not. In such a case, the callback is usually ignored. If
+      // the handler happened to both call the callback and complete a returned Promise, whichever happens first will
+      // win and the latter will be ignored.
+      const wrappedCallback = plugin._wrapCallback(callback, span);
+      const maybePromise = safeExecuteInTheMiddle(
+        () => original.apply(this, [event, context, wrappedCallback]),
+        error => {
+          if (error != null) {
+            // Exception thrown synchronously before resolving callback / promise.
+            plugin._endSpan(span, error, () => {});
+          }
+        }
+      ) as Promise<{}> | undefined;
+      if (
+        typeof maybePromise !== 'undefined' &&
+        typeof maybePromise.then === 'function'
+      ) {
+        return maybePromise.then(
+          value =>
+            new Promise(resolve =>
+              plugin._endSpan(span, undefined, () => resolve(value))
+            ),
+          (err: Error | string) =>
+            new Promise((resolve, reject) =>
+              plugin._endSpan(span, err, () => reject(err))
+            )
+        );
+      }
+      return maybePromise;
+    };
+  }
+
+  setTracerProvider(tracerProvider: TracerProvider) {
+    super.setTracerProvider(tracerProvider);
+    this._tracerProvider = tracerProvider;
+  }
+
+  private _wrapCallback(original: Callback, span: Span): Callback {
+    const plugin = this;
+    return function wrappedCallback(this: {}, err, res) {
+      diag.debug('executing wrapped lookup callback function');
+
+      plugin._endSpan(span, err, () => {
+        diag.debug('executing original lookup callback function');
+        return original.apply(this, [err, res]);
+      });
+    };
+  }
+
+  private _endSpan(
+    span: Span,
+    err: string | Error | null | undefined,
+    callback: () => void
+  ) {
+    if (err) {
+      span.recordException(err);
+    }
+
+    let errMessage;
+    if (typeof err === 'string') {
+      errMessage = err;
+    } else if (err) {
+      errMessage = err.message;
+    }
+    if (errMessage) {
+      span.setStatus({
+        code: SpanStatusCode.ERROR,
+        message: errMessage,
+      });
+    }
+
+    span.end();
+    // Lambda only runs on Node so always true in practice.
+    if (this._tracerProvider instanceof NodeTracerProvider) {
```

Comment:
I don't think we need to be that strict. Someone could implement it's own `TracerProvider` by extending `BasicTracerProvider`.